unit/java_bytecode/generate_concrete_generic_type/generate_java_generic_type.cpp

```diff
@@ -0,0 +1,216 @@
+        {
+          generic_components_counter++;
+        }
+      }
+      // Our class had 2 generic fields that should be recognised.
+      REQUIRE(generic_components_counter==2);
+    }
+  }
+
+}
+
+
+SCENARIO(
+  "generate_java_generic_type_from_file_two_params",
+  "[CORE][java_bytecode][generate_java_generic_type][!mayfail]")
+{
+  std::unique_ptr<languaget> java_lang(new_java_bytecode_language());
+
+  // Configure the path loading
+  cmdlinet command_line;
+  command_line.set(
+    "java-cp-include-files",
+    "./java_bytecode/generate_concrete_generic_type");
+  config.java.classpath.emplace_back(
+    "./java_bytecode/generate_concrete_generic_type");
+
+  // Configure the language
+  null_message_handlert message_handler;
+  java_lang->get_language_options(command_line);
+  java_lang->set_message_handler(message_handler);
+
+  std::istringstream java_code_stream("ignored");
+
+  GIVEN("A generic java type with two generic parameters, like a Hashtable")
+  {
+    java_lang->parse(java_code_stream, "generic_two_parameters.class");
+
+    symbol_tablet new_symbol_table;
+    java_lang->typecheck(new_symbol_table, "");
+
+    java_lang->final(new_symbol_table);
+
+    REQUIRE(new_symbol_table.has_symbol(
+      "java::generic_two_parameters$KeyValuePair"));
+    THEN("The class should have two subtypes in the vector of the types of "
+           "the generic components.")
+    {
+      const symbolt &class_symbol=new_symbol_table.lookup(
+        "java::generic_two_parameters$KeyValuePair");
+      const typet &symbol_type=class_symbol.type;
+
+      REQUIRE(symbol_type.id()==ID_struct);
+      class_typet class_type=to_class_type(symbol_type);
+      REQUIRE(class_type.is_class());
+
+      REQUIRE(class_type.subtypes().size()==2);
+    }
+  }
+
+}
+
+
+SCENARIO(
+  "generate_java_generic_type_from_file_uninstantiated_param",
+  "[CORE][java_bytecode][generate_java_generic_type][!shouldfail]")
+{
+  std::unique_ptr<languaget> java_lang(new_java_bytecode_language());
+
+  // Configure the path loading
+  cmdlinet command_line;
+  command_line.set(
+    "java-cp-include-files",
+    "./java_bytecode/generate_concrete_generic_type");
+  config.java.classpath.emplace_back(
+    "./java_bytecode/generate_concrete_generic_type");
+
+  // Configure the language
+  null_message_handlert message_handler;
+  java_lang->get_language_options(command_line);
+  java_lang->set_message_handler(message_handler);
+
+  std::istringstream java_code_stream("ignored");
+
+  // This test should fail.
+  GIVEN("A generic java type with a field that refers to another generic with"
+          " an uninstantiated parameter.")
+  {
+    java_lang->parse(java_code_stream, "generic_unknown_field.class");
+
+    symbol_tablet new_symbol_table;
+    java_lang->typecheck(new_symbol_table, "");
+
+    java_lang->final(new_symbol_table);
+
+    REQUIRE(new_symbol_table.has_symbol
+      ("java::generic_unknown_field$element<T>"));
```

Comment:
I think this is the opposite of what we should have? Specifically we shouldn't create a new type representing `element<T>` (we already have the symbol `element` to represent this).